### PR TITLE
Typed annotations

### LINKED ISSUE #20: Add first class support for attribute syntax?
Maybe add first class support for the various forms of attributes syntaxes instead of taking them as plain text.

Seems closely related to #18, if we end up using `impl SrcCode`? We might use something like `Annotation` enum for the various types, (doc, attribute, etc)

### PR DESCRIPTION
**Pull request by milesgranger**
Will close #20

### CODE REVIEW

**Comment by DarkDrek** [comment]
I don't think we should mix documentation and annotations.
And naming wise it shouldn't be ModuleAttr and Atrr but Scope-Attribute and Item-Attribute (or something like that).
Because `#![warn(...)]` can also be the fist thing inside a function and then it applies to the whole function scope.

**Comment by milesgranger** [comment]
Ah right, of course. Thank you. 👍

File: src/gen/annotation.rs [hunk 1]
```diff
@@ -0,0 +1,49 @@
+//!
+//! Create `enum` objects
```
Comment (DarkDrek) [inline src/gen/annotation.rs]: Shouldn't this be this "create annotation objects/items"?

File: src/gen/annotation.rs [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+/// Represents a single Rust annotation to a module, function, etc.
+#[derive(Debug, Clone, Serialize, Deserialize)]
+pub enum Annotation {
+    /// Annotation representing an attribute for an item. ie. `#[foo(bar)]`
+    ItemAttr(String),
+    /// Annotation representing a scoped level attribute. ie. `#![warn(...)]`
+    ScopeAttr(String),
+}
+
+impl<S: ToString> From<S> for Annotation {
```
Comment (DarkDrek) [inline src/gen/annotation.rs]: How about TryFrom here? https://doc.rust-lang.org/std/convert/trait.TryFrom.html

File: src/gen/module.rs [hunk 3]
```diff
@@ -123,11 +123,11 @@ impl Module {
...
-    fn inner_annotations(&mut self) -> &mut Vec<String> {
+    fn inner_annotations(&mut self) -> &mut Vec<Annotation> {
         &mut self.inner_annotations
     }
 
-    fn outer_annotations(&mut self) -> &mut Vec<String> {
+    fn outer_annotations(&mut self) -> &mut Vec<Annotation> {
         &mut self.outer_annotations
     }
 }
```
Comment (DarkDrek) [inline src/gen/module.rs]: With this change we do not need to have inner and outer anymore. Inner can only be scope Annotations and outer just item Annotations.

File: tests/utilities/mod.rs [hunk 4]
```diff
@@ -11,6 +11,7 @@ pub trait Verify: SrcCode {
 
     fn generate_and_verify(&self) -> String {
         let src_code = self.generate();
+        println!("{}", &src_code);
```
Comment (DarkDrek) [inline tests/utilities/mod.rs]: Can your remove the println(src_code) from the other test that use generate_and_verify?

**Comment by DarkDrek** [comment]
Can you implement the Verify trait for Annotation?

File: src/gen/annotation.rs (hunk 2 above)
Comment (milesgranger) [reply src/gen/annotation.rs]: hmm. That's a good idea, but I'm getting this:
```
  --> src/gen/annotation.rs:30:1
   |
30 | impl<S: ToString> TryFrom<S> for Annotation {
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = note: conflicting implementation in crate `core`:
           - impl<T, U> std::convert::TryFrom<U> for T
             where U: std::convert::Into<T>;
```
Seem related to this: https://github.com/rust-lang/rust/issues/50133
Any ideas?

**Comment by milesgranger** [comment]
Not sure if it's possible? Doesn't appear `Parse` is implemented for `Attribute` https://docs.rs/syn/1.0.7/syn/struct.Attribute.html

So I get this:
```
  --> tests/utilities/mod.rs:19:6
   |
19 | impl Verify for Annotation {
   |      ^^^^^^ the trait `syn::parse::Parse` is not implemented for `syn::attr::Attribute`
```

**Comment by DarkDrek** [comment]
Can you apply this https://docs.rs/syn/1.0.5/syn/struct.Attribute.html#parsing-from-tokens-to-attribute ? But for this would have to provide a new trait or a supertrait of Verify so it may be too much work

File: src/gen/annotation.rs (hunk 2 above)
Comment (DarkDrek) [reply src/gen/annotation.rs]: Ohhh this sucks 😞 Can you add a comment linking to this issue so we can fix this when its fixed.

**Comment by milesgranger** [comment]
Awesome, I can look at it tomorrow probably. Also, I would not be offended if you feel like hijacking this PR if you want. :+1:

**Comment by DarkDrek** [comment]
Ok if I find time I can try 👍. I hope you also didn't fell offended by my remarks 👋

**Comment by milesgranger** [comment]
haha, never! More comments the better, my Rust skills are a work in progress anyway so I appreciate it!

**Comment by milesgranger** [comment]
If you're good with it, we can add a separate issue for implementing `Verify`/`Parse` for `Annotation` after this is in.

**Comment by milesgranger** [comment]
..in the larger picture, should this rather be called `Attribute` instead of `Annotation`?

**Pull request merged**